org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java

```diff
@@ -715,46 +887,170 @@ public static IJavaElement findElementAtSelection(ITypeRoot unit, int line, int
+			if (contents != null && !contents.isBlank()) {
 				IDocument document = new Document(contents);
+				ICompilationUnit workingCopy = null;
 				try {
 					offset = document.getLineOffset(line) + column;
 					if (offset > -1) {
-						String name = parse(contents, offset);
-						if (name == null) {
+						if (offset > 0 && !isJavaIdentifierOrPeriod(contents.charAt(offset))) {
+							offset = offset - 1;
+						}
+						if (monitor != null && monitor.isCanceled()) {
+							return null;
+						}
+						IRegion region = parseRegion(contents, offset);
+						if (region == null) {
 							return null;
 						}
-						SearchPattern pattern = SearchPattern.createPattern(name, IJavaSearchConstants.TYPE,
-								IJavaSearchConstants.DECLARATIONS, SearchPattern.R_FULL_MATCH);
-
-						IJavaSearchScope scope = createSearchScope(unit.getJavaProject(), preferenceManager);
-
-						List<IJavaElement> elements = new ArrayList<>();
-						SearchRequestor requestor = new SearchRequestor() {
-							@Override
-							public void acceptSearchMatch(SearchMatch match) {
-								if (match.getElement() instanceof IJavaElement) {
-									elements.add((IJavaElement) match.getElement());
-								}
+						if (monitor != null && monitor.isCanceled()) {
+							return null;
+						}
+						workingCopy = unit.getWorkingCopy(new WorkingCopyOwner() {
+						}, monitor);
+						workingCopy.getBuffer().setContents(contents);
+						workingCopy.becomeWorkingCopy(monitor);
+						workingCopy.makeConsistent(monitor);
+						IJavaElement[] sourceElements = workingCopy.codeSelect(region.getOffset(), region.getLength());
+						if (sourceElements == null || sourceElements.length == 0) {
+							return null;
+						}
+						IJavaElement element;
+						if (sourceElements.length != 1) {
+							// they could be package fragments.
+							// We need to select the one that matches the package fragment of the current unit
+							IPackageFragment packageFragment = (IPackageFragment) unit.getParent();
+							IJavaElement found = Stream.of(sourceElements).filter(e -> e.equals(packageFragment)).findFirst().orElse(null);
+							if (found == null) {
+								// this would be a binary package fragment
+								element = sourceElements[0];
+							} else {
+								element = found;
 							}
-						};
-						SearchEngine searchEngine = new SearchEngine();
-						searchEngine.search(pattern,
-								new SearchParticipant[] { SearchEngine.getDefaultSearchParticipant() }, scope,
-								requestor, null);
+						} else {
+							element = sourceElements[0];
+						}
+						String name = contents.substring(region.getOffset(), region.getOffset() + region.getLength());
+						List<IJavaElement> elements = search(unit, name, element, preferenceManager, monitor);
+						if (monitor != null && monitor.isCanceled()) {
+							return null;
+						}
+						if (elements.isEmpty()) {
+							elements = search(unit.getJavaProject(), name, element, preferenceManager, monitor);
+						}

```

Comment:
The codeSelect above already returns a JavaElement, it it possible to make other places to use this result directly? I didn't get the point to use search engine to search the element again.

Reply:
The codeselect finds a java element in a ICompilationUnit we have created using JDTUtils.getWorkingCopy(). We can't return a java element from ICompilationUnit which is created temporarily and destroyed. We have to search a class file.

Reply:
> We can't return a java element from ICompilationUnit which is created temporarily and destroyed.

If this is the only concern, then you can have the outermost caller to create the workingcopy, and destroy the workingcopy until the LSP response returns.

We need to search three times to find the currently selected element, which seems a bit too complicated.

Reply:
I have updated the PR. Could, you, please, review it.